### PR TITLE
Added a date prop to FormatDate and RelativeTime

### PR DESCRIPTION
**Pull request by YakimaProgrammer**
Adds support for passing either a JavaScript date or a String to the Shoelace FormatDate and RelativeTime components as per Shoelace's custom-elements.json. Before, these components rendered times relative to when they were instantiated, with no way to set a date through Laminar. Now, any date can be specified.
Here's an example of the prop in action:
```scala
p("Using a js.Date, this should be, '3 days ago': ", sl.RelativeTime.of(
    _.date(new js.Date(js.Date.now() - (24*60*60*1000 * 3)))
))
p("Using a String, this should be, '7/27/2024': ", sl.FormatDate.of(
    _.date("2024-07-28T02:41:45.682Z")
))
```
I'd love to hear what you think of this approach to union types!

### CODE REVIEW

**Comment by raquo** [comment]
Thanks for the PR! This is definitely better than no support at all previously, and I would say that in general, this is a very reasonable approach. It's relatively simple to implement, requiring few changes. Its main weakness is that the users are left to deal with `js.Date | String` union type, which is not ideal, in two ways (OTOH):

1. When **reading** the property, we want to read `js.Date`. I _assume_ that reading `.ref.date` from e.g. a Shoelace RelativeTime element will always give you a `js.Date`, not a `String`, but the typings for this prop produced by this PR give users `js.Date | String` ([here](https://github.com/raquo/laminar-shoelace-components/pull/4/files#diff-c3e51fc5951c261ac172a9c3c381cb7566d1b9c089fd3dd0621277225f41e823R82)), which is less specific than it could be.

    I see potential ways to improve this, but it seems like too much work, for the benefit that it would provide. _Writing_ props in Laminar apps is needed more often than _reading_ them. This can be improved later / separately.

2. Could you please check in your branch that the following usages of the `<--` method compile?

```scala
p("Using a js.Date, this should be, '3 days ago': ", sl.RelativeTime.of(
    _.date <-- Val(new js.Date(js.Date.now() - (24*60*60*1000 * 3)))
))
val strSource = Val("2024-07-28T02:41:45.682Z") // Source[String] type inferred here
p("Using a String, this should be, '7/27/2024': ", sl.FormatDate.of(
    _.date <-- strSource
))
```

Basically, the `<--` method expects a `Source[js.Data | String]`, but does it compile if we give it e.g. `Source[String]`? `Source[+A]` is covariant in A, so I would hope so. If there is a problem with the `<--` method, this probably needs a solution with an implicit conversion from `Source[A]` to `Source[A | *]` (and `Source[* | A]`) at Laminar or Airstream level, or something like that.

Either way, this PR looks good to me, and I'm good to merge it – just note the one trivial code review comment. Thanks again!

**Comment by YakimaProgrammer** [comment]
Thank you so much for taking a look! I'd like to apologize for my slow response time.  
I totally agree with you on that first point there that it makes since to have `js.Date` be the dominate type when reading to clean things up for the user - I have an idea for how we could make that happen and I'll send it up soon. I'll also give those test cases a look as well - thanks for pointing those out!

**Comment by YakimaProgrammer** [comment]
I've been experimenting a bit more with `.ref` to see what the non-union type options might look like to make `.ref.date` a little easier to read out of. As of now, it will give back the object and type of whatever the property was last set to, not necessarily only `js.Date`.
```scala
// `String`, prints "2024-08-03 "
println(sl.FormatDate.of(_.date("2024-08-03 ")).ref.date)
// `js.Date`, prints "Sat Aug 03 2024 00:00:00 GMT-0700 (Pacific Daylight Time)"
println(sl.FormatDate.of(_.date(new js.Date("2024-08-03 "))).ref.date)
```
I tried a couple of `js.Date`-only options on the JavaScript side, but I couldn't come up with anything that didn't feel clunky or unsafe. I ultimately found that keeping `.ref.date` as a union and pattern matching might be the most straightforward, if less than perfect, approach:
```scala
val dateSource = Val("2024-08-03 ")
val myComp = sl.FormatDate.of(
  _.date <-- dateSource
)
val date = myComp.ref.date match {
  case d: js.Date => d
  case d: String => new js.Date(d)
}
println(date.toDateString()) //"Sat Aug 03 2024"
```
In particular, I tried some modifications at the facade level to rewrite to a `js.Date` on access, but the solutions seemed cumbersome and difficult to fit back into the existing code generation framework. I also considered removing the union at the facade level in favor of only `js.Date`, given that Laminar controls the component and access to it. However, I'm not entirely certain this would cut off all potential code paths (in Scala or JavaScript) where the underlying property might be set to a `String`, potentially leading to unexpected errors and behaviors.

While `Source[js.Date]` and `Source[String]` are both implicitly promoted to `Source[js.Date | String]` for `sl.FormatDate.of(_.date <-- someSource)` out of the box, I also I also considered a `dateProp` that would only accept `js.Date`s, but I'm unsure whether this would provide significant benefits beyond serving as a reminder to validate your data before passing it further. Moreover, this alone would not change the typing expressed by `.ref.date`.
```scala
protected def dateProp(name: String): HtmlProp[js.Date, js.Date | String] = L.htmlProp(name, new Codec[js.Date, js.Date | String] {
  override def decode(domValue: js.Date | String): js.Date = domValue match {
    case d: js.Date => d
    case d: String => new js.Date(d)
  }

  override def encode(scalaValue: js.Date): js.Date | String = scalaValue
});

// This no longer works if this `dateProp` implementation is used instead of `asIsProp`
sl.FormatDate.of(
  _.date <-- Val("2024-08-03 ")
)
```

Regarding the test cases, with the current `asIsProp` exposing the union directly in Laminar, the examples using `Source[js.Date]` and `Source[String]` were implicitly promoted to the union type without issue, allowing users to use the two types interchangeably as expected out of the box.

Ultimately, I'll happily follow your lead with the decisions you make regarding the handling of union types. Also, thank you for the discussion and consideration around union types - I think this is an interesting use case and I appreciate your thoughts on how we should approach it.

**Comment by raquo** [comment]
Thanks for looking at the options! Overall I concur.

> As of now, it will give back the object and type of whatever the property was last set to, not necessarily only js.Date.

Ah, that's surprising. I assumed Shoelace would normalize the property value, but no, it just uses whatever value you provide whenever it needs to render ([source](https://github.com/shoelace-style/shoelace/blob/next/src/components/format-date/format-date.component.ts#L56)), and so if you ask for the property value, you get whatever you put in.

So my wish to narrow the types was based on a bad assumption that Shoelace narrowed them. They don't, so we don't have to either, the union type is fine. This project is not yet at a stage where we can worry about providing a better API for Shoelace, simply mirroring the original API is a good start.

So, I think your original encoding of the union type is perfectly fine in these circumstances. The usages I was worried about work great, so all is good. So I'm just gonna go ahead and merge this.

--

Going forward, with props that have union types, I think following your pattern for `date` is definitely fine if the underlying Shoelace prop behaves the same way as this `date` prop, i.e. when you read it, you get what you wrote to it, without it being normalized into any one type.

If you stumble upon a union type prop that does normalize its contents, i.e. when you go to read it, it's always in one of the types regardless of what you put in, this pattern is probably ok too, but please take a note of that prop's behaviour to mark it for future improvement.

Thanks for your work on this!

**Comment by YakimaProgrammer** [comment]
That sounds great! I think that a lot of these cases will likely follow a pattern similar to `.date`, but obviously we'll take everything on a case-by-case basis. Thanks again for taking a look and for your thoughts while we considered this case!
